integration/spark-interfaces-scala/README.md

```diff
@@ -0,0 +1,208 @@
+`OutputDatasetWithDelegate` have been created. They allow assigning facets to a dataset, while
+still letting other code to extract metadata for the same dataset. The classes contain `node` object
+property which defines node within logical plan to contain more metadata about the dataset. 
+
+An example implementation for `ReplaceIcebergData` node:
+
+```scala
+override def getOutputs(context: OpenLineageContext): List[OutputDatasetWithFacets] = {
+    if (!table.isInstanceOf[DataSourceV2Relation]) {
+      List()
+    } else {
+      val relation = table.asInstanceOf[DataSourceV2Relation]
+      val datasetFacetsBuilder: DatasetFacetsBuilder = {
+        new OpenLineage.DatasetFacetsBuilder()
+          .lifecycleStateChange(
+          context
+            .openLineage
+            .newLifecycleStateChangeDatasetFacet(
+              OpenLineage.LifecycleStateChangeDatasetFacet.LifecycleStateChange.OVERWRITE,
+              null
+            )
+        )
+      }
+      
+      // enrich dataset with additional facets like a dataset version
+      DatasetVersionUtils.getVersionOf(relation) match {
+        case Some(version) => datasetFacetsBuilder.version(
+          context
+            .openLineage
+            .newDatasetVersionDatasetFacet(version)
+        )
+        case None =>
+      }
+
+      // return output dataset while pointing that more dataset details shall be extracted from
+      // `relation` object.
+      List(
+        OutputDatasetWithDelegate(
+          relation,
+          datasetFacetsBuilder,
+          new OpenLineage.OutputDatasetOutputFacetsBuilder()
+        )
+      )
+    }
+  }
+```
+ 
+## Run and Job Facets
+
+The library also allows enriching OpenLineage events being sent by any job and run facets. 
+
+`io.openlineage.spark.builtin.nodes.FacetRegistry` object is a static singleton which allows
+registering arbitrary facets to be emitted within the next OpenLineage event. 
+
+This can be done with methods:
+ * `def registerRunFacet(name: String, facet: RunFacet): Unit`
+ * `def registerJobFacet(name: String, facet: JobFacet): Unit`
+
+When initializing OpenLineage, facet registry gets `OpenLineageContext` which can be used to generate
+the facets. It contains `OpenLineage` object with builders for all the facets defined in the 
+specification.
+
+Please note that library gives no guarantee if the next event is going to be sent. Registering a 
+facet after the job has finished will result in facets not being sent.
+
+## Column Level Lineage
+
+Lineage extraction is based on optimized logical plan which is a tree with a root being output
+dataset and leafs which are input datasets. In order to track column level lineage, not only input
+and outputs fields need to be tracked, but also dependencies between.
+
+Each node within plan has to understand which input attributes does it consume and how do they 
+affect output attributes produced by the node. Attribute fields within plan are identified by
```

Comment:
Each node within plan has to understand which input attributes it consumes and how they affect output attributes produced by the node.